### PR TITLE
Enable SimilarityMaps in C++

### PR DESCRIPTION
**Pull request by greglandrum**
This allows the generation of similarity maps from C++ using the RDKit's MolDraw2D objects. 
It adds backend functionality for contouring data and rendering data on a grid that will certainly be useful for other purposes.

I'm not quite done polishing this, but I would love feedback on what the API looks like and what pieces might be missing.

Here's an example of the output:
image

The code used to produce that:
```
mol = Chem.MolFromSmiles('COc1cccc2cc(C(=O)NCCCCN3CCN(c4cccc5nccnc54)CC3)oc21')
refmol = Chem.MolFromSmiles('CCCN(CCCCN1CCN(c2ccccc2OC)CC1)Cc1ccc2ccccc2c1')
d = Draw.MolDraw2DCairo(400, 400)
d.ClearDrawing()
_, maxWeight = sm.GetSimilarityMapForFingerprint(
  refmol, mol, lambda m, i: sm.GetMorganFingerprint(m, i, radius=2, fpType='bv'), draw2d=d)
d.FinishDrawing()
with open('similarityMap1_out.png', 'w+') as outf:
  outf.write(d.GetDrawingText())```

### CODE REVIEW

File: Code/GraphMol/MolDraw2D/MolDraw2D.h [hunk 1]
```diff
@@ -29,14 +29,39 @@
...
+           fabs(b - other.b) <= tol &&
+           (ignoreAlpha || fabs(a - other.a) <= tol);
+  };
+  DrawColour operator+(const DrawColour &other) const {
+    return DrawColour(r + other.r, g + other.g, b + other.b, a + other.a);
+  }
+  DrawColour operator-(const DrawColour &other) const {
+    return DrawColour(r - other.r, g - other.g, b - other.b, a - other.a);
+  }
+  DrawColour operator/(double v) const {
```
Comment (bp-kelley) [inline Code/GraphMol/MolDraw2D/MolDraw2D.h]: v!=0 precondition?

File: Code/GraphMol/MolDraw2D/MolDraw2DQt.cpp [hunk 2]
```diff
@@ -27,8 +27,7 @@ MolDraw2DQt::MolDraw2DQt(int width, int height, QPainter &qp, int panelWidth,
 // ****************************************************************************
 void MolDraw2DQt::setColour(const DrawColour &col) {
   MolDraw2D::setColour(col);
-  QColor this_col(int(255.0 * col.get<0>()), int(255.0 * col.get<1>()),
-                  int(255.0 * col.get<2>()));
+  QColor this_col(int(255.0 * col.r), int(255.0 * col.g), int(255.0 * col.b));
```
Comment (bp-kelley) [inline Code/GraphMol/MolDraw2D/MolDraw2DQt.cpp]: I'm assuming that MolDraw2Qt drops the alpha channel?

File: Code/GraphMol/MolDraw2D/MolDraw2D.h [hunk 3]
```diff
@@ -29,14 +29,39 @@
...
+  bool operator==(const DrawColour &other) const {
+    return r == other.r && g == other.g && b == other.b && a == other.a;
+  }
+  bool feq(const DrawColour &other, double tol = 0.001,
+           bool ignoreAlpha = true) const {
+    return fabs(r - other.r) <= tol && fabs(g - other.g) <= tol &&
+           fabs(b - other.b) <= tol &&
+           (ignoreAlpha || fabs(a - other.a) <= tol);
+  };
+  DrawColour operator+(const DrawColour &other) const {
```
Comment (bp-kelley) [inline Code/GraphMol/MolDraw2D/MolDraw2D.h]: I'm a bit bothered by this as r,g,b,a could end up > 1.

File: Code/Numerics/Conrec.h [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#include <vector>
+#include <Geometry/point.h>
+
+namespace conrec {
+struct ConrecSegment {
```
Comment (bp-kelley) [inline Code/Numerics/Conrec.h]: Neat!

File: Code/Numerics/Conrec.h [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+  PRECONDITION(nc > 0, "no contours");
+  PRECONDITION(iub > ilb, "bad bounds");
+  PRECONDITION(jub > jlb, "bad bounds");
+
+#define xsect(p1, p2) (h[p2] * xh[p1] - h[p1] * xh[p2]) / (h[p2] - h[p1])
+#define ysect(p1, p2) (h[p2] * yh[p1] - h[p1] * yh[p2]) / (h[p2] - h[p1])
+
+  int m1, m2, m3, case_value;
+  double dmin, dmax, x1 = 0, x2 = 0, y1 = 0, y2 = 0;
+  int i, j, m;
```
Comment (bp-kelley) [inline Code/Numerics/Conrec.h]: So FORTRANY!

File: Code/Numerics/testConrec.cpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    std::vector<conrec::ConrecSegment> segs;
+    const size_t nContours = 10;
+    double isoLevels[nContours];
+    for (size_t i = 0; i < nContours; ++i) {
+      isoLevels[i] = (i + 1) * (maxV / (nContours + 1));
+    }
+    conrec::Contour(grid, 0, gridSz - 1, 0, gridSz - 1, xps, yps, nContours,
+                    isoLevels, segs);
+
+    std::ofstream outs("./blah.svg");
```
Comment (bp-kelley) [inline Code/Numerics/testConrec.cpp]: I'm not quite sure what this test does, is "blah.svg" supposed to be compared to something?

File: rdkit/Chem/Draw/UnitTestSimilarityMaps.py [hunk 7]
```diff
@@ -165,6 +166,18 @@ def testSimilarityMapKWArgs(self):
     # chiral center drops:
     self.assertTrue(weights[2] > weights2[2])
 
+  def testSimilarityMapsMolDraw2D(self):
+    # nothing really sensible to test here, just make sure things run
```
Comment (bp-kelley) [inline rdkit/Chem/Draw/UnitTestSimilarityMaps.py]: Ok, I understand the above test now.

File: Code/GraphMol/MolDraw2D/MolDraw2DUtils.h [hunk 8]
```diff
@@ -75,6 +75,32 @@ RDKIT_MOLDRAW2D_EXPORT void updateDrawerParamsFromJSON(MolDraw2D &drawer,
...
+      {0.557, 0.004, 0.322, 0.5},
+      {1, 1, 1, 0.5},
+      {0.153, 0.392, 0.098, 0.5}};  // similarity map color scheme
+};
+RDKIT_MOLDRAW2D_EXPORT void contourAndDrawGrid(
+    MolDraw2D &drawer, const double *grid, const std::vector<double> &xcoords,
+    const std::vector<double> &ycoords, size_t nContours,
+    std::vector<double> &levels, const ContourParams &ps = ContourParams());
+
+RDKIT_MOLDRAW2D_EXPORT void contourAndDrawGaussians(
```
Comment (bp-kelley) [inline Code/GraphMol/MolDraw2D/MolDraw2DUtils.h]: It looks like one can supply nContours OR levels.  I find having them in the same function is confusing.  I might have two functions:

```
void contourAndDrawGaussians( 
   MolDraw2D &drawer, const std::vector<Point2D> &locs,
    const std::vector<double> &heights, const std::vector<double> &widths,
    size_t nContours, 
    const ContourParams &ps = ContourParams());
```

```
void contourAndDrawGaussians( 
   MolDraw2D &drawer, const std::vector<Point2D> &locs,
    const std::vector<double> &heights, const std::vector<double> &widths,
    const std::vector<double> &widths,
    const ContourParams &ps = ContourParams());
```

**Comment by bp-kelley** [comment]
I would love to make the contours "clickable".  I can see a lot of utility here.  Perhaps using an alpha and raising them above the molecule..

File: Code/GraphMol/MolDraw2D/MolDraw2D.h (hunk 3 above)
Comment (greglandrum) [reply Code/GraphMol/MolDraw2D/MolDraw2D.h]: Could very well happen when doing average colors (i.e. `(c1+c2)/2`).
I think this kind of problem is best handled when the color is actually used

File: Code/Numerics/Conrec.h (hunk 5 above)
Comment (greglandrum) [reply Code/Numerics/Conrec.h]: yep
there was no way I was going to rewrite that code though.

File: Code/GraphMol/MolDraw2D/MolDraw2DUtils.h (hunk 8 above)
Comment (greglandrum) [reply Code/GraphMol/MolDraw2D/MolDraw2DUtils.h]: The idea behind having both was to allow the code to return the contour values being used if they are automatically selected.
I will add some documentation about this as well as some sanity checking in the code.

File: Code/GraphMol/MolDraw2D/MolDraw2D.h (hunk 1 above)
Comment (greglandrum) [reply Code/GraphMol/MolDraw2D/MolDraw2D.h]: yep

File: Code/GraphMol/MolDraw2D/MolDraw2DQt.cpp (hunk 2 above)
Comment (greglandrum) [reply Code/GraphMol/MolDraw2D/MolDraw2DQt.cpp]: was an oversight since I don't use the Qt thing

**Comment by greglandrum** [comment]
I think the API for this isn't too horrible. We can at least give it a try for a while and see what we think.
I'd be willing to merge now. @bp-kelley, what do you think?

**Comment by bp-kelley** [comment]
Any reason not to make a default for:

const std::vector<double> &levels=const std::vector<double>()

Otherwise I’m happy to merge

**Comment by greglandrum** [comment]
Ok, made that suggested change and added a call to contour plain grids.

**Pull request merged**